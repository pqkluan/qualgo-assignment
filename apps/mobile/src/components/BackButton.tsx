import { useNavigation } from '@react-navigation/native';
import { FC } from 'react';
import { TouchableOpacity, View } from 'react-native';
import { createStyleSheet, useStyles } from 'react-native-unistyles';

import ChevronLeftIcon from '@mobile/assets/icons/chevron-left.svg';

type Props = {
	onPress?: () => void;
};

export const BackButton: FC<Props> = (props) => {
	const { onPress } = props;

	const navigation = useNavigation();

	const { styles, theme } = useStyles(stylesheet);

	return (
		<TouchableOpacity testID='back-button' onPress={onPress ?? navigation.goBack}>
			<View style={styles.backButton}>
				<ChevronLeftIcon
					testID='chevron-left-icon'
					width={20}
					height={20}
					color={theme.colors.background}
				/>
			</View>
		</TouchableOpacity>
	);
};

export const BackButtonHeight = 42;

const stylesheet = createStyleSheet((theme, rt) => ({
	header: {
		position: 'absolute',
		left: theme.margins.xl,
		paddingTop: rt.insets.top,
	},
	backButton: {
		justifyContent: 'center',
		alignItems: 'center',

		height: BackButtonHeight,
		width: BackButtonHeight,
		borderRadius: BackButtonHeight / 2,

		backgroundColor: theme.colors.typography,
	},
}));
