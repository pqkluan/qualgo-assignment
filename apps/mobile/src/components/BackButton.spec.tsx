import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';

import { BackButton } from './BackButton';

const mockedGoBack = jest.fn();

jest.mock('@react-navigation/native', () => ({
	useNavigation: () => ({ goBack: mockedGoBack }),
}));

describe('BackButton', () => {
	it('match snapshot', () => {
		const view = render(<BackButton />);
		expect(view).toMatchSnapshot();
	});

	it('renders button icon', () => {
		const view = render(<BackButton />);

		expect(view.getByTestId('chevron-left-icon')).toBeDefined();
	});

	it('trigger navigation.goBack upon press by default', () => {
		mockedGoBack.mockClear();

		const view = render(<BackButton />);
		fireEvent.press(view.getByTestId('back-button'));

		expect(mockedGoBack).toHaveBeenCalledTimes(1);
	});

	it('trigger onPress props upon press', () => {
		const onPress = jest.fn();

		const view = render(<BackButton onPress={onPress} />);
		fireEvent.press(view.getByTestId('back-button'));

		expect(onPress).toHaveBeenCalledTimes(1);
	});
});
