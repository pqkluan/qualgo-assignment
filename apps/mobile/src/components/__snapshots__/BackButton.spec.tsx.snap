// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackButton match snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="back-button"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#000000",
        "borderRadius": 21,
        "height": 42,
        "justifyContent": "center",
        "width": 42,
      }
    }
  >
    <SvgMock
      color="#FFFFFF"
      height={20}
      testID="chevron-left-icon"
      width={20}
    />
  </View>
</View>
`;
